File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        
        unordered_map<int, int> mp;
        int left = 0, right = 0, ans = 0;
        
        while(left < n and right < n){
            if(mp.find(s[right]) != mp.end()){
                left = max(left, mp[s[right]] + 1);
            }
            
            mp[s[right]] = right;
            int currLen = right - left + 1;
            ans = max(ans, currLen);
            right++;
        }
        
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        occurences = {}
        
        left, right, ans = 0, 0, 0
        n = len(s)
        
        while left < n and right < n:
            if s[right] in occurences:
                left = max(left, occurences[s[right]] + 1)
            
            occurences[s[right]] = right
            ans = max(ans, right - left + 1)
            right += 1
        return ans